### PR TITLE
Move ibis/sql module to ibis/backends/base_sqlalchemy

### PR DESCRIPTION
**Pull request by renato2099**
I just moved/renamed `ibis/sql` into `ibis/backends/base_sqlalchemy` but I am not sure how to test this, so I might have missed something (sorry in advanced!).
I will check around maybe I just missed something.

### CODE REVIEW

File: ibis/backends/base_sqlalchemy/compiler.py
Comment (datapythonista) [inline ibis/backends/base_sqlalchemy/compiler.py:792]: I don't think this should be changed.

File: ibis/backends/base_sqlalchemy/mysql/api.py
Comment (datapythonista) [inline ibis/backends/base_sqlalchemy/mysql/api.py:3]: I think I mentioned in the issue, that better to wait for #2446 to be merged. It's ok to get started, but don't move the mysql backend, since it's move there.

Also, you are changing these imports in a way that doesn't make sense, there are three different modules and you're making them the same?

File: ibis/backends/base_sqlalchemy/mysql/api.py
Comment (datapythonista) [inline ibis/backends/base_sqlalchemy/mysql/api.py:19]: This is the sqalchemy module (a third-party package), which obviously didn't change.

File: ibis/backends/base_sqlalchemy/mysql/compiler.py
Comment (datapythonista) [inline ibis/backends/base_sqlalchemy/mysql/compiler.py:11]: More imports wrongly changed.

File: ibis/backends/clickhouse/tests/test_client.py
Comment (datapythonista) [inline ibis/backends/clickhouse/tests/test_client.py:65]: This is a configuration option named `sql`, not a module name, not sure if you automatically changed `sql.` by `base_sqlalchemy.`, but many changes don't seem to make sense.

**Review by datapythonista — CHANGES_REQUESTED** [review]
Thanks @renato2099. Lots of issues, I added some comments that should help you understand what's not correct, but didn't review in detail. Probably worth to start again from scratch. With the feedback I gave you, you should be able to get a much better PR.

After opening the PR, you need to wait (like 30 minutes) and then check the results of the continuous integration. If tests or anything is broken, you'll get a red build, and the logs will tell you what's wrong. Probably worth checking the contributing page on how to test Ibis locally, and try to get the tests for sqlite and pandas backends working before you open a PR.

**Comment by datapythonista** [comment]
@renato2099, #2446 has been merged now, do you want to move forward with this PR?

File: ibis/backends/clickhouse/tests/test_client.py
Comment (renato2099) [reply ibis/backends/clickhouse/tests/test_client.py]: yeah, something like that, sorry for the noise!

**Comment by renato2099** [comment]
hey @datapythonista , yeah I think I had misunderstood the actual task, and went ahead and moved a bunch of code from one place to another and didn't actually check (sorry!).
So I was more careful this time, I only moved `ibis/sql/alchemy.py` -> `ibis/backends/base_sqlalchemy/alchemy.py` but I am wondering now from your original description:

> would be good to move the SQLAlchemy shared code from ibis/sql to ibis/backends/base_sqlalchemy

are you also referring to files such as compiler.py and transforms.py?

Also I keep on getting failures in CI, any suggestions how to debug them? I am not sure what they mean (https://github.com/ibis-project/ibis/pull/2468/checks?check_run_id=1271969429#step:4:559) thanks again!

```
FAILED ibis/tests/all/test_client.py::test_load_data_sqlalchemy[Csv] - Attrib...
FAILED ibis/tests/all/test_client.py::test_load_data_sqlalchemy[Pandas] - Att...
FAILED ibis/tests/all/test_client.py::test_load_data_sqlalchemy[Parquet] - At...
```

**Comment by datapythonista** [comment]
Looks like you still have all the mysql files in the diff. You may want to merge master into your branch.

For what to move, it was a bit tricky, but the problem was that the root of `sql/` had the files shared among backends, and then the backends themselves (`mysql`, `postgres`,...) were also in this same directory. Now this is not the case anymore, since all backends have been moved, including `mysql`, which was there when you started this PR.

So now it's much simpler. It's just moving `ibis/sql` to `ibis/backends/base_sqlalchemy`. This includes all the files in the directory:
```
sql/compiler.py
sql/transforms.py
sql/alchemy.py
sql/__init__.py
```
That's trivial, the work in this PR is updating the imports of the stuff that you're moving. Most of the imports will be from `ibis/backends/{mysql|postgres|sqlite}`, but it's likely that some tests and some other stuff are also importing this. Some greps and running the tests and see what fails should help find all the broken imports.

I hope it's clearer now, sorry for not providing enough context, and thanks for working on this @renato2099.

**Comment by jreback** [comment]
@datapythonista how is this related to base_sql?

**Comment by datapythonista** [comment]
> @datapythonista how is this related to base_sql?

You will know better, but my understanding is that `ibis/sql` has the code shared among backends based on SQLAlchemy, so renaming it to `backends/base_sqlalchemy` makes sense. And `backends/base_sql` is the code of the custom compiler implemented initially for Impala.

I think both are unrelated, they are the base for different kinds of backends, depending if they are implemented using a SQLAlchemy engine or from scratch.

Does this make sense to you, or did I misunderstood anything?

**Comment by jreback** [comment]
@datapythonista that sounds right. Yeah we should document this in the ``__init__`` of each backend what it is supposed to cover (and of course to share code if possible). ok if they are orthogonal that makes it simpler.

**Review by jreback — APPROVED** [review]
lgtm @datapythonista merge when ready

**Review by datapythonista — APPROVED** [review]
Thanks @renato2099, great job.

As said, if you can move the other files in `ibis/sql` to `ibis/backends/base_sqlalchemy` in a follow up PR, that would be great.

**Comment by renato2099** [comment]
hey @datapythonista , thanks for the comments and the guidance 👍 
I will move the rest of the fiiles under `sql` later on today, so should the `ibis/sql` folder be removed?

**Comment by datapythonista** [comment]
> hey @datapythonista , thanks for the comments and the guidance 
> I will move the rest of the fiiles under `sql` later on today, so should the `ibis/sql` folder be removed?

Thanks  @renato2099. Yes, the `ibis/sql` directory should be empty/removed after this refactoring, since it was the location of the SQLAlchemy backends, which now will be spread across `backends/base_sqlalchemy`, `backends/mysql`, `backends/sqlite` and `backends/postgres`.

**Pull request merged**
